tzrec/datasets/dataset.py

```diff
@@ -317,6 +321,92 @@ def _build_batch(self, input_data: Dict[str, pa.Array]) -> Batch:
                 input_data = _expand_tdm_sample(
                     input_data, pos_sampled, neg_sampled, self._data_config
                 )
+            elif self._enable_hstu:
+                seq_attr = self._sampler._item_id_field
+                if pa.types.is_string(input_data[seq_attr].type):
+                    input_data_k_split = pc.split_pattern(
+                        input_data[seq_attr], self._seq_str_delim
+                    )
+                    # Only keep the target item for previous sequence
+                    # example: [1,2,3,4,5] -> [2,3,4,5] (target item list)
+                    input_data_k_split_slice = pc.list_flatten(
+                        pc.list_slice(input_data_k_split, start=1)
+                    )
+                    pre_seq = pc.list_flatten(input_data_k_split).to_numpy(
+                        zero_copy_only=False
+                    )
+                    # Replace the last item with '-1'
+                    pre_seq[input_data_k_split.offsets.to_numpy()[1:] - 1] = "-1"
+                    mask = pre_seq != "-1"
+                    # Only keep the training sequence, example: [1,2,3,4,5] -> [1,2,3,4]
+                    pre_seq_filter = pre_seq[mask]
+                    pre_seq_filter_offsets = pa.array(
+                        np.concatenate(
+                            [
+                                np.array([0]),
+                                input_data_k_split.offsets[1:].to_numpy(
+                                    zero_copy_only=False
+                                )
+                                - np.arange(
+                                    1,
+                                    len(
+                                        input_data_k_split.offsets[1:].to_numpy(
+                                            zero_copy_only=False
+                                        )
+                                    )
+                                    + 1,
+                                ),
+                            ]
+                        )
+                    )
+                    pre_seq_filter_reshaped = pa.ListArray.from_arrays(
+                        pre_seq_filter_offsets, pre_seq_filter
+                    )
+                    pre_seq_filter_reshaped_joined = pc.binary_join(
+                        pre_seq_filter_reshaped, self._seq_str_delim
+                    )
+                if self._mode == Mode.TRAIN:
+                    # Training using all possible target items
+                    input_data[seq_attr] = input_data_k_split_slice
+                elif self._mode == Mode.EVAL:
+                    # Evaluation using the last item for previous sequence
+                    input_data[seq_attr] = input_data_k_split.values.take(
+                        pa.array(input_data_k_split.offsets.to_numpy()[1:] - 1)
+                    )
+                sampled = self._sampler.get(input_data)
+                for k, v in sampled.items():
```

Comment:
Make line 377-409 a function and move it to datasets/utils.py. Add comments to explain the logic and add a unit test for the function.